README
Readme for OpenTitus
Git version
http://opentitus.sourceforge.net/

OpenTitus is released under the Gnu GPL version 3 or (at your option) any later version of the license, and it is released "as is"; without any warranty.

The OpenTitus team:
Eirik Stople - Main developer - eirik@pcfood.net
Eitan Tal - Valuable assistant, technical advisor

Thanks to:
Eric Zmiro and his team in Titus Interactive who made the great games Titus the Fox and Moktar!
Jesse, who made a fansite for the original game and an editor and provided me information how the original game files was built up, and assisted me with the makefile and MinGW!
The SDL project (LibSDL, LibSDL_mixer)!
The GNU project (GCC, GDB and GPL)!
MinGW/msys!
Ubuntu (programming environment)!
Linux!
Dosbox!
Dropbox!
GEdit!
Notepad++!
Google search!
Chocolate-Doom's great OPL (adlib) emulator!
Various tutorials/information sources on the net!


Git instructions:
Compile according to COMPILING.
To test out your build, copy it into the game folder and follow the instructions inside bin/README.
Precompiled binaries are available in builds/


You need the original game files to make use of OpenTitus. OpenTitus parses the original files. It works with both Titus the Fox and Moktar.
Place the original game files in the "titus" or "moktar" folder. The files must be placed according to titus.conf.

Runtime dependencies:
SDL runtime library
SDL_mixer runtime library

To run the game, run opentitus (in Linux) or opentitus.exe (windows).

To quickly start up Titus, copy a Titus game directory to the same folder as the opentitus executable, and name the folder "titus".
To quickly start up Moktar, copy a Moktar game directory to the same folder as the opentitus executable, and name the folder "moktar". You can modify titus.conf if you like, or use the preconfigured "titus_moktar.conf" which you need to rename "titus.conf".

The game engine are ported from the original game, and modified to fit OpenTitus' level structure. Movement are equal to the original game.

If you can find some bugs or differences between OpenTitus and the original games, feel free to contact us!

Enjoy!



"Titus the Fox: To Marrakech and Back" (1992) and "Lagaf': Les Aventures de Moktar - Vol 1: La Zoubida" (1991)
was developed by, and is probably copyrighted by Titus Software, which, according to Wikipedia, stopped buisness in 2005.
OpenTitus is not affiliated with Titus Software.

The source code have been developed during several years, and some parts of the code are old. This code doesn't follow the more recent style, such as "unsigned char" instead of "uint8". As the old code still works, it's not that important to fix it.

The configuration information below are old, and there may be faults in this information. The preconfigured config files should work.
The levelviewer and the game is configured by a file in the same directory; "titus.conf". This file is parsed line by line, and lines starting with # will be skipped.
Keywords:
reswidth = Number of pixels wide, default 320 (originally 640, but the "pixels" in the original game was two pixels wide, so 320 should give the correct size)
resheight = Number of pixels height, default 200 (originally 400 (intro) or 384 (in-game), 200 should give the correct size)
bitdepth = Number of bits pr. pixel, default 32 (the game data files have a 16 colour palette, which means 4 bpp, but higher bpp will work as well)
ingamewidth = Numbers og pixels visible in-game, default 320.
ingameheight = Numbers og pixels visible in-game, default 192.
videomode = Video mode; 0 is window mode and 1 is fullscreen, default 0
levelcount = Number of levels, default 15 (15 in Titus the Fox and 16 in Moktar)
level 1-15 (1-16 for Moktar) = Location of the level files for all levels, default LEVELx.SQZ (the order is "0J123456789BCEFG" for Moktar, Titus is equal except it doesn't have level F)
sprites = Location of the sprite file, default SPREXP.SQZ (SPRITES.SQZ in Moktar)
game = The game opentitus will parse; 0 for Titus the Fox and 1 for Moktar. This is necessary because of different sprite sizes, different menus and different level titles.
logo = The company logo file, default TITUS.SQZ
intro = The game intro image file, default TITRE.SQZ
menu = The menu, default MENU.SQZ
finish = The image displayed when game is completed, default LEVELA.SQZ (this doesn't exist in Moktar)
*format = Image format. 0: 16-colour grayscale, 1: 16-colour (ordinary colours), 2: 256 colours + palette. Logo, intro and menu are 256 colour, *EGA files are 16-colour, and LEVELA.SQZ are grayscale
font = Font file, default FONTS.SQZ

Music spesific:
moduleintro = Location of the module (.mod, amiga audio) file containing the intro music
moduleprelevel = Location of the module file containing the pre-level music
modulegameover = Location of the module file containing the game over music
modulelevelfile 1-6 = location of the module files containing the level music (number 6 is the wedding "level" music)
module*loop = Loop information for the modules. Nonnegative values indicates the location to continue playback when the module is played. -1 indicates to play the module only once.
modulelevelfileloop 1-6 = Should be self-explainable
modulelevel 1-15 (1-16 for Moktar): Defines what modulelevelfile to use on each level


docs/CHANGELOG
0.8.0
Not released yet
This version contains a working adlib emulator (from the chocolate-doom project), and audio is 
implemented. Various bugfixes.

0.7.0
23.10.11
This version contains a complete game engine, and should be almost identical to the original game.
This version doesn't support audio.

0.6.0
15.09.11
This snapshot is based on 0.4.0 (C code). What's new is a working game engine, made after deep
analysis of TTF's original game engine. The engine does appearently handle player, objects, gates
and elevators flawlessly. This snapshot is tested mostly with Titus, and it should handle Moktar as
well. What's remaining is the enemies, and the adlib audio.

0.5.0
Engine rewritten in C++, discontinued.

0.4.0
01.03.09
This is the first audio-enabled version. It can play amiga modules by now, by using LibMikMod or
SDL-mixer. I haven't looked at Atari and Amstrad yet. Adlib audio is difficult to achieve, because
a good adlib emulator is required. The best emulator by now is the MAME adlib emulator, but that is
covered by the MAME license, and thus makes it difficult to include in OpenTitus. Audio is
pauseable by using F3. LibMikMod can play music in the same process (which reduces the framerate to
8FPS), or it can play in a separate process (recommanded). SDL-mixer will also play music in a
separate process. It is possible to do some basic compile-time configuration by manually editing
config.h. I have added a very basic movement engine. This is the last version programmed in C. The
next will be C++. This version is also possible to load in NetBeans.

0.3.0
24.12.08
This version successfully loads the level in a TITUS_level structure and displays the levels by
reading from this structure. It is also possible to move titus "noclip" style to examine the levels.
What's left is animation and the movement engine, and music. The "levelviewer" executable is not
compatible with the new functions, and is simply skipped in the makefile. This will be the first
version on Sourceforge. This version also contains Jesses' documentations.

0.2.1
29.11.08
Jesse modified the makefile to compile all, and he confirmed that the project still compiled in
MinGW in windows, by using the makefile.

0.2
27.11.08
A snapshot. This is not a separate version because of an achieved goal, but to show the progress.
What's new is the "opentitus" executable. The source have a better structure now, and the main
functions should be easier to understand and debug. I have written a new function that extracts
the font used in the game, and correctly prints it on the screen. After some playing with alpha
(or transparency) settings and keyboard handling routines, I have successfully cloned the intro of
the game and the menu. To be specific: Intro text, titus logo, game image, menu, password prompting
and the pre-level screen. This game is now also fully functional with Moktar (by modifying the
titus.conf setting file). This snapshot also includes a level structure, which should make the game
quite configurable (extendable enemy-, object-, gate-, elevator- and bonus records count, and
extendable level width and height, and possibility to use custom sprites and background tiles with
custom size and bit depth). What the program misses now is the game engine and music. I haven't
made anything new regarding the level viewer, except cleaning the source file by moving functions
in separate source files. As usual, I have only compiled the source in Linux, so I can't guarantee
any windows functionality; however, Jesse told me that version 0.1 did compile successfully in MinGW.

0.1
16.11.08
First "official" version. The project isn't released to the public yet.
This version includes a program that extracts SQZ files, and a level viewer. This is not interactive;
only the upper left part of the level is displayed (read prior 0.1). The project also have a simple
config file, with possibilities to change resolution and video mode, and the location of the original
data files. This version is only compatible with Titus the Fox (Moktar doesn't work yet) in a Linux
environment.

Prior 0.1
22.09.08 - 15.11.08
The project starts with a function that extracts the SQZ files. Functions that extracts background
tiles, bonuses, elevators, enemies, objects and Titus himself is made, one by one, and finally this
is displayed on the screen using SDL, which results in version 0.1.


src/draw.c
/*   
 * Copyright (C) 2008 - 2011 The OpenTitus team
 *
 * Authors:
 * Eirik Stople
 *
 * "Titus the Fox: To Marrakech and Back" (1992) and
 * "Lagaf': Les Aventures de Moktar - Vol 1: La Zoubida" (1991)
 * was developed by, and is probably copyrighted by Titus Software,
 * which, according to Wikipedia, stopped buisness in 2005.
 *
 * OpenTitus is not affiliated with Titus Software.
 *
 * OpenTitus is  free software; you can redistribute  it and/or modify
 * it under the  terms of the GNU General  Public License as published
 * by the Free  Software Foundation; either version 3  of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT  ANY  WARRANTY;  without   even  the  implied  warranty  of
 * MERCHANTABILITY or  FITNESS FOR A PARTICULAR PURPOSE.   See the GNU
 * General Public License for more details.
 */

/* draw.c
 * Draw functions
 *
 * Global functions:
 * int TFR_SCREENM(): Draw tiles on the backbuffer (copy from the tile screen)
 * int viewstatus(TITUS_level *level, bool countbonus): View status screen (F4)
 * int flip_screen(bool slow): Flips the screen and a short delay
 * int INIT_SCREENM(TITUS_level *level): Initialize backbuffer
 * int DISPLAY_COUNT(TITUS_level *level): Draw energy
 * int fadeout(): Fade the screen to black
 * int view_password(TITUS_level *level, uint8 level_index): Display the password
 */

#include <stdio.h>
#include <stdlib.h>
#include "SDL2/SDL.h"
#include "globals.h"
#include "backbuffer.h"
#include "sprites.h"
#include "draw.h"
#include "settings.h"
#include "common.h"
#include "tituserror.h"
#include "original.h"

SDL_Surface *sprite_from_cache(TITUS_level *level, TITUS_sprite *spr);

int TFR_SCREENM() { //Draw tiles on the backbuffer (copy from the tile screen)
    SDL_Rect src, dest;

    //First of all: make the screen black, at least the lower part of the screen
    dest.x = 0;
    dest.y = screen_height * 16;
    dest.w = screen_width * 16;
    dest.h = 200 - screen_height * 16;
    SDL_FillRect(screen, &dest, SDL_MapRGB(screen->format, 0, 0, 0));

    // Tile screen:  | Output screen:
    //               |
    // D | C         | A | B
    // -   -         | -   -
    // B | A         | C | D
    //
    // The screens are splitted in 4 parts by BITMAP_XM and BITMAP_YM
    // The code below will move the 4 rectangles with tiles to their right place on the output screen


    //Upper left on screen (A)
    src.x = BITMAP_XM * 16;
    src.y = BITMAP_YM * 16;
    src.w = (screen_width - BITMAP_XM) * 16;
    src.h = (screen_height - BITMAP_YM) * 16;
    dest.x = 0;
    dest.y = 0;
    dest.w = src.w;
    dest.h = src.h;
    SDL_BlitSurface(tilescreen, &src, screen, &dest);

    //Upper right on screen (B)
    src.x = 0;
    src.y = BITMAP_YM * 16;
    src.w = BITMAP_XM * 16;
    src.h = (screen_height - BITMAP_YM) * 16;
    dest.x = (screen_width - BITMAP_XM) * 16;
    dest.y = 0;
    dest.w = src.w;
    dest.h = src.h;
    SDL_BlitSurface(tilescreen, &src, screen, &dest);

    //Lower left on screen (C)
    src.x = BITMAP_XM * 16;
    src.y = 0;
    src.w = (screen_width - BITMAP_XM) * 16;
    src.h = BITMAP_YM * 16;
    dest.x = 0;
    dest.y = (screen_height - BITMAP_YM) * 16;
    dest.w = src.w;
    dest.h = src.h;
    SDL_BlitSurface(tilescreen, &src, screen, &dest);

    //Lower right on screen (D)
    src.x = 0;
    src.y = 0;
    src.w = BITMAP_XM * 16;
    src.h = BITMAP_YM * 16;
    dest.x = (screen_width - BITMAP_XM) * 16;
    dest.y = (screen_height - BITMAP_YM) * 16;
    dest.w = src.w;
    dest.h = src.h;
    SDL_BlitSurface(tilescreen, &src, screen, &dest);
}


//Loop through all sprites, and draw the sprites that should be visible on the screen (NOT by using the visible flag, it uses the coordinates)
//If the flash bit is set, the first 3 planes will be 0, the last plane will be normal (colour & 0x01, odd colors gets white, even colours gets black)


DISPLAY_SPRITES(TITUS_level *level) {
    int16 i;
    char buffer[7]; //xxx ms

    for (i = level->elevatorcount - 1; i >= 0; i--) {
        display_sprite(level, &(level->elevator[i].sprite));
    }

    for (i = level->trashcount - 1; i >= 0; i--) {
        display_sprite(level, &(level->trash[i]));
    }

    for (i = level->enemycount - 1; i >= 0; i--) {
        display_sprite(level, &(level->enemy[i].sprite));
    }

    for (i = level->objectcount - 1; i >= 0; i--) {
        display_sprite(level, &(level->object[i].sprite));
    }

    display_sprite(level, &(level->player.sprite3));
    display_sprite(level, &(level->player.sprite2));
    display_sprite(level, &(level->player.sprite));

    if (GODMODE) {
        SDL_Print_Text("GODMODE", 30 * 8, 0 * 12);
    }
    if (NOCLIP) {
        SDL_Print_Text("NOCLIP", 30 * 8, 1 * 12);
    }

#ifdef DEBUG_VERSION
    if (DISPLAYLOOPTIME) {
        sprintf(buffer, "%3u ms", LOOPTIME);
        SDL_Print_Text(buffer, 30 * 8, 2 * 12); //Loop time in ms

        sprintf(buffer, "FPS %u", FPS_LAST);
        SDL_Print_Text(buffer, 30 * 8, 4 * 12); //Last second's FPS count

        sprintf(buffer, "CL %d", LAST_CLOCK);
        SDL_Print_Text(buffer, 30 * 8, 6 * 12); //Clock

        sprintf(buffer, "CORR %d", LAST_CLOCK_CORR);
        SDL_Print_Text(buffer, 30 * 8, 8 * 12); //Correction to the clock


        for (i = 0; i <= 15; i++) {
            sprintf(buffer, "%d %3u", i, SUBTIME[i]);
            SDL_Print_Text(buffer, 0 * 8, i * 12); //Sub times from main loop in ms
        }

        sprintf(buffer, "%d %3u", i, SUBTIME[i]);
        SDL_Print_Text(buffer, 0 * 8, i * 12);
    }

#endif

}

display_sprite(TITUS_level *level, TITUS_sprite *spr) {
    SDL_Surface *image;
    SDL_Rect src, dest;
    if (!spr->enabled) {
        return;
    }
    if (spr->invisible) {
        return;
    }
    spr->visible = false;
    //At this point, the buffer should be the correct size

    if (!spr->flipped) {
        dest.x = spr->x - spr->spritedata->refwidth - (BITMAP_X << 4);
    } else {
        dest.x = spr->x + spr->spritedata->refwidth - spr->spritedata->data->w - (BITMAP_X << 4);
    }
    dest.y = spr->y + spr->spritedata->refheight - spr->spritedata->data->h + 1 - (BITMAP_Y << 4);

    if ((dest.x >= screen_width * 16) || //Right for the screen
      (dest.x + spr->spritedata->data->w < 0) || //Left for the screen
      (dest.y + spr->spritedata->data->h < 0) || //Above the screen
      (dest.y >= screen_height * 16)) { //Below the screen
        return;
    }

    image = sprite_from_cache(level, spr);
/*
    if ((spr->flipped != spr->flipped_last) ||
      (spr->flash != spr->flash_last)) {
        SDL_FreeSurface(spr->buffer);
        spr->buffer = copysurface(spr->spritedata->data, spr->flipped, spr->flash);
    }

    spr->flipped_last = spr->flipped;
    spr->flash_last = spr->flash;

    SDL_Surface *image = spr->buffer;
*/
    src.x = 0;
    src.y = 0;
    src.w = image->w;
    src.h = image->h;

    if (dest.x < 0) {
        src.x = 0 - dest.x;
        src.w -= src.x;
        dest.x = 0;
    }
    if (dest.y < 0) {
        src.y = 0 - dest.y;
        src.h -= src.y;
        dest.y = 0;
    }
    if (dest.x + src.w > screen_width * 16) {
        src.w = screen_width * 16 - dest.x;
    }
    if (dest.y + src.h > screen_height * 16) {
        src.h = screen_height * 16 - dest.y;
    }

    SDL_BlitSurface(image, &src, screen, &dest);



    spr->visible = true;
    spr->flash = false;

}

SDL_Surface *sprite_from_cache(TITUS_level *level, TITUS_sprite *spr) {
    TITUS_spritecache *cache = level->spritecache;
    TITUS_spritedata *spritedata = level->spritedata[spr->number];
    TITUS_spritebuffer *spritebuffer;
    uint8 index;
    int16 i;

    if (spr->flipped) {index = 1;} else {index = 0;};

    if (spr->flash) {
        for (i = cache->count - cache->tmpcount; i < cache->count; i++) {
            spritebuffer = cache->spritebuffer[i];
            if (spritebuffer != NULL) {
                if ((spritebuffer->spritedata == spritedata) &&
                  (spritebuffer->index == index + 2)) {
                    return spritebuffer->data; //Already in buffer
                }
            }
        }
        //Not found, load into buffer
        cache->cycle2++;
        if (cache->cycle2 >= cache->count) { //The last 3 buffer surfaces is temporary (reserved for flash)
            cache->cycle2 = cache->count - cache->tmpcount;
        }
        spritebuffer = cache->spritebuffer[cache->cycle2];
        SDL_FreeSurface(spritebuffer->data); //Free old surface
        spritebuffer->data = copysurface(spritedata->data, spr->flipped, spr->flash);
        spritebuffer->spritedata = spritedata;
        spritebuffer->index = index + 2;
        return spritebuffer->data;
    } else {
        if (spritedata->spritebuffer[index] == NULL) {
            cache->cycle++;
            if (cache->cycle + cache->tmpcount >= cache->count) { //The last 3 buffer surfaces is temporary (reserved for flash)
                cache->cycle = 0;
            }
            spritebuffer = cache->spritebuffer[cache->cycle];
            if (spritebuffer->spritedata != NULL) {
                spritebuffer->spritedata->spritebuffer[spritebuffer->index] = NULL; //Remove old link
            }
            SDL_FreeSurface(spritebuffer->data); //Free old surface
            spritebuffer->data = copysurface(spritedata->data, spr->flipped, spr->flash);
            spritebuffer->spritedata = spritedata;
            spritebuffer->index = index;
            spritedata->spritebuffer[index] = spritebuffer;
        }
        return spritedata->spritebuffer[index]->data;
    }
}


int flip_screen(bool slow) {
    int tick = SDL_GetTicks();
    SDL_Flip(screen);
    int oldtick = tick;
    tick = SDL_GetTicks();
    SUBTIME[14] = tick - oldtick;

    //if (slow) {
        NO_FAST_CPU(slow); //TODO: 
    //}
}

/*
NO_FAST_CPU(bool slow) {
    int tick, duration, delay, tick2;
    tick = SDL_GetTicks();
    if (slow) {
        delay = 29; //28.53612, fps: 70.09Hz/2
    } else {
        delay = 10;
    }
    LOOPTIME = (tick - LAST_CLOCK);
    delay = delay - (tick - LAST_CLOCK);
    if ((delay < 0) || (delay > 40)) {
        delay = 1;
    }
    SDL_Delay(delay);
    //do {
        //SDL_Delay(1);
    //    tick = SDL_GetTicks();
    //    duration = abs(LAST_CLOCK - tick);
    //} while (duration < delay);
    tick2 = SDL_GetTicks();
    if ((tick2 / 1000) != (LAST_CLOCK / 1000)) {
        FPS_LAST = FPS;
        FPS = 0;
    }
    FPS++;
    LAST_CLOCK_CORR = tick2 - tick - delay;

    LAST_CLOCK = tick2;
    SUBTIME[15] = LAST_CLOCK - tick;
}
*/

NO_FAST_CPU(bool slow) {
    int tick, duration, delay, tick2;
    tick = SDL_GetTicks();
    if (slow) {
        delay = 29; //28.53612, fps: 70.09Hz/2
    } else {
        delay = 10;
    }
    LOOPTIME = (tick - LAST_CLOCK);
    delay = delay - (tick - LAST_CLOCK) - LAST_CLOCK_CORR;
    LAST_CLOCK_CORR = 0;
    if (delay > 40) {
        delay = 1;
    } else if (delay < 0) {
        LAST_CLOCK_CORR = (0 - delay) / 2; // To reduce LAST_CLOCK_CORR
        delay = 0;
    }



    tick2 = SDL_GetTicks();
    duration = abs(tick - tick2);
    while (duration < delay) {
        titus_sleep();
        //SDL_Delay(1);
        tick2 = SDL_GetTicks();
        duration = abs(tick - tick2);
    }



    //SDL_Delay(delay);




    //do {
        //SDL_Delay(1);
    //    tick = SDL_GetTicks();
    //    duration = abs(LAST_CLOCK - tick);
    //} while (duration < delay);
    tick2 = SDL_GetTicks();
    if ((tick2 / 1000) != (LAST_CLOCK / 1000)) {
        FPS_LAST = FPS;
        FPS = 0;
    }
    FPS++;


    LAST_CLOCK_CORR += tick2 - tick - delay;
        if (LAST_CLOCK_CORR > 25) {
            LAST_CLOCK_CORR = 25;
        }

    LAST_CLOCK = tick2;

    SUBTIME[15] = LAST_CLOCK - tick;
}


int viewstatus(TITUS_level *level, bool countbonus){
    int retval, i, j;
    char tmpchars[10];
    SDL_FillRect(screen, NULL, SDL_MapRGB(screen->format, 0, 0, 0));

    if (game == 0) { //Titus
        SDL_Print_Text("LEVEL", 13 * 8, 12 * 5);
        SDL_Print_Text("EXTRA BONUS", 10 * 8, 10 * 12);
        SDL_Print_Text("LIVES", 10 * 8, 11 * 12);
    } else if (game == 1) { //Moktar
        SDL_Print_Text("ETAPE", 13 * 8, 12 * 5);
        SDL_Print_Text("EXTRA BONUS", 10 * 8, 10 * 12);
        SDL_Print_Text("VIE", 10 * 8, 11 * 12);
    }


    sprintf(tmpchars, "%d", level->levelnumber + 1);
    SDL_Print_Text(tmpchars, 25 * 8 - strlen(tmpchars) * 8, 12 * 5);

    SDL_Print_Text(leveltitle[level->levelnumber], 0, 12 * 5 + 16);
    sprintf(tmpchars, "%d", level->extrabonus);
    SDL_Print_Text(tmpchars, 28 * 8 - strlen(tmpchars) * 8, 10 * 12);

    sprintf(tmpchars, "%d", level->lives);
    SDL_Print_Text(tmpchars, 28 * 8 - strlen(tmpchars) * 8, 11 * 12);

    SDL_Flip(screen);

    if (countbonus && (level->extrabonus >= 10)) {
        retval = waitforbutton();
        if (retval < 0) {
            return retval;
        }
        while (level->extrabonus >= 10) {
            for (i = 0; i < 10; i++) {
                level->extrabonus--;
                sprintf(tmpchars, "%2d", level->extrabonus);
                SDL_Print_Text(tmpchars, 28 * 8 - strlen(tmpchars) * 8, 10 * 12);
                SDL_Flip(screen);
                for (j = 0; j < 15; j++) {
                    NO_FAST_CPU(false);
                }
            }
            level->lives++;
            sprintf(tmpchars, "%d", level->lives);
            SDL_Print_Text(tmpchars, 28 * 8 - strlen(tmpchars) * 8, 11 * 12);
            SDL_Flip(screen);
            for (j = 0; j < 10; j++) {
                NO_FAST_CPU(false);
            }
        }
    }

    retval = waitforbutton();
    if (retval < 0)
        return retval;

    SDL_FillRect(screen, NULL, SDL_MapRGB(screen->format, 0, 0, 0));
    SDL_Flip(screen);

    return (0);
}

int INIT_SCREENM(TITUS_level *level) {
    CLOSE_SCREEN();
    BITMAP_X = 0;
    BITMAP_Y = 0;
    BITMAP_XM = 0;
    BITMAP_YM = 0;
    uint8 i, j;
    for (i = 0; i < 20; i++) {
        for (j = 0; j < 12; j++) {
            DISPLAY_CHAR(level, level->tilemap[BITMAP_Y + j][BITMAP_X + i], j, i);
        }
    }
    do {
        scroll(level);
    } while (YSCROLL_CENTER || XSCROLL_CENTER);
    OPEN_SCREEN();
}


int DISPLAY_COUNT(TITUS_level *level) {
    subto0(&(BAR_FLAG));
    if (BAR_FLAG != 0) {
        DISPLAY_ENERGY(level);
    }
}


void DISPLAY_ENERGY(TITUS_level *level) {
    uint8 offset = 96;
    uint8 i;
    SDL_Rect dest;
    for (i = 0; i < level->player.hp; i++) { //Draw big bars (4px*16px, spacing 4px)
        dest.x = offset;
        dest.y = 9;
        dest.w = 4;
        dest.h = 16;
        SDL_FillRect(screen, &dest, SDL_MapRGB(screen->format, 255, 255, 255));
        offset += 8;
    }
    for (i = 0; i < MAXIMUM_ENERGY - level->player.hp; i++) { //Draw small bars (4px*4px, spacing 4px)
        dest.x = offset;
        dest.y = 15;
        dest.w = 4;
        dest.h = 4;
        SDL_FillRect(screen, &dest, SDL_MapRGB(screen->format, 255, 255, 255));
        offset += 8;
    }
}

int fadeout() {
    SDL_Surface *image;
    int activedelay = 1;
    SDL_Event event;
    unsigned int fade_time = 1000;
    unsigned int tick_start = 0;
    unsigned int image_alpha = 0;
    SDL_Rect src, dest;

    src.x = 0;
    src.y = 0;
    src.w = screen->w;
    src.h = screen->h;

    dest.x = 0;
    dest.y = 0;
    dest.w = screen->w;
    dest.h = screen->h;

    image = SDL_ConvertSurface(screen, screen->format, SDL_SWSURFACE);

    SDL_BlitSurface(screen, &src, image, &dest);
    tick_start = SDL_GetTicks();
    while (image_alpha < 255) //Fade to black
    {
        if (SDL_PollEvent(&event)) {
            if (event.type == SDL_QUIT) {
                SDL_FreeSurface(image);
                return (-1);
            }

            if (event.type == SDL_KEYDOWN) {
                if (event.key.keysym.scancode == SDL_SCANCODE_ESCAPE) {
                    SDL_FreeSurface(image);
                    return (-1);
                }
                if (event.key.keysym.scancode == KEY_MUSIC) {
                    AUDIOMODE++;
                    if (AUDIOMODE > 1) {
                        AUDIOMODE = 0;
                    }
                    if (AUDIOMODE == 1) {
                        startmusic();
                    }
                }
            }
        }

        image_alpha = (SDL_GetTicks() - tick_start) * 256 / fade_time;

        if (image_alpha > 255)
            image_alpha = 255;

        SDL_SetSurfaceAlphaMod(image, 255 - image_alpha);
        SDL_FillRect(screen, NULL, SDL_MapRGB(screen->format, 0, 0, 0));
        SDL_BlitSurface(image, &src, screen, &dest);
        SDL_Flip(screen);

        titus_sleep();
    }
    SDL_FreeSurface(image);

}

int view_password(TITUS_level *level, uint8 level_index) {
    //Display the password !
    char tmpchars[10];
    int retval;

    CLOSE_SCREEN();
    SDL_FillRect(screen, NULL, SDL_MapRGB(screen->format, 0, 0, 0));
    SDL_Flip(screen);

    if (game == 0) { //Titus
        SDL_Print_Text("LEVEL", 13 * 8, 13 * 8);
    } else if (game == 1) { //Moktar
        SDL_Print_Text("ETAPE", 13 * 8, 13 * 8);
    }
    sprintf(tmpchars, "%d", level_index + 1);
    SDL_Print_Text(tmpchars, 25 * 8 - strlen(tmpchars) * 8, 13 * 8);

    SDL_Print_Text("CODE", 14 * 8, 10 * 8);
    SDL_Print_Text(levelcode[level_index], 20 * 8, 10 * 8);

    SDL_Flip(screen);
    retval = waitforbutton();
    if (retval < 0)
        return retval;

    //SDL_Flip(screen);
    OPEN_SCREEN();
    return (0);
}

int loadpixelformat(SDL_PixelFormat **pixelformat){
    int i;

    *pixelformat = (SDL_PixelFormat *)SDL_malloc(sizeof(SDL_PixelFormat));
    if (*pixelformat == NULL) {
        sprintf(lasterror, "Error: Not enough memory to initialize palette!\n");
        return (TITUS_ERROR_NOT_ENOUGH_MEMORY);
    }

    (*pixelformat)->palette = (SDL_Palette *)SDL_malloc(sizeof(SDL_Palette));
    if ((*pixelformat)->palette == NULL) {
        sprintf(lasterror, "Error: Not enough memory to initialize palette!\n");
        return (TITUS_ERROR_NOT_ENOUGH_MEMORY);
    }

    (*pixelformat)->palette->ncolors = 16;

    (*pixelformat)->palette->colors = (SDL_Color *)SDL_malloc(sizeof(SDL_Color) * (*pixelformat)->palette->ncolors);
    if ((*pixelformat)->palette->colors == NULL) {
        sprintf(lasterror, "Error: Not enough memory to initialize palette!\n");
        return (TITUS_ERROR_NOT_ENOUGH_MEMORY);
    }

    for (i = 0; i < (*pixelformat)->palette->ncolors; i++) {
        (*pixelformat)->palette->colors[i].r = orig_palette_colour[i].r;
        (*pixelformat)->palette->colors[i].g = orig_palette_colour[i].g;
        (*pixelformat)->palette->colors[i].b = orig_palette_colour[i].b;
    }

    (*pixelformat)->BitsPerPixel = 8;
    (*pixelformat)->BytesPerPixel = 1;

    (*pixelformat)->Rloss = 0;
    (*pixelformat)->Gloss = 0;
    (*pixelformat)->Bloss = 0;
    (*pixelformat)->Aloss = 0;

    (*pixelformat)->Rshift = 0;
    (*pixelformat)->Gshift = 0;
    (*pixelformat)->Bshift = 0;
    (*pixelformat)->Ashift = 0;

    (*pixelformat)->Rmask = 0;
    (*pixelformat)->Gmask = 0;
    (*pixelformat)->Bmask = 0;
    (*pixelformat)->Amask = 0;

    //(*pixelformat)->colorkey = 0;
    //(*pixelformat)->alpha = 255;

    return (0);
}

int loadpixelformat_font(SDL_PixelFormat **pixelformat){
    int i;

    *pixelformat = (SDL_PixelFormat *)SDL_malloc(sizeof(SDL_PixelFormat));
    if (*pixelformat == NULL) {
        sprintf(lasterror, "Error: Not enough memory to initialize palette!\n");
        return (TITUS_ERROR_NOT_ENOUGH_MEMORY);
    }

    (*pixelformat)->palette = (SDL_Palette *)SDL_malloc(sizeof(SDL_Palette));
    if ((*pixelformat)->palette == NULL) {
        sprintf(lasterror, "Error: Not enough memory to initialize palette!\n");
        return (TITUS_ERROR_NOT_ENOUGH_MEMORY);
    }

    (*pixelformat)->palette->ncolors = 16;

    (*pixelformat)->palette->colors = (SDL_Color *)SDL_malloc(sizeof(SDL_Color) * (*pixelformat)->palette->ncolors);
    if ((*pixelformat)->palette->colors == NULL) {
        sprintf(lasterror, "Error: Not enough memory to initialize palette!\n");
        return (TITUS_ERROR_NOT_ENOUGH_MEMORY);
    }

    for (i = 0; i < (*pixelformat)->palette->ncolors; i++) {
        (*pixelformat)->palette->colors[i].r = orig_palette_font_colour[i].r;
        (*pixelformat)->palette->colors[i].g = orig_palette_font_colour[i].g;
        (*pixelformat)->palette->colors[i].b = orig_palette_font_colour[i].b;
    }

    (*pixelformat)->BitsPerPixel = 8;
    (*pixelformat)->BytesPerPixel = 1;

    (*pixelformat)->Rloss = 0;
    (*pixelformat)->Gloss = 0;
    (*pixelformat)->Bloss = 0;
    (*pixelformat)->Aloss = 0;

    (*pixelformat)->Rshift = 0;
    (*pixelformat)->Gshift = 0;
    (*pixelformat)->Bshift = 0;
    (*pixelformat)->Ashift = 0;

    (*pixelformat)->Rmask = 0;
    (*pixelformat)->Gmask = 0;
    (*pixelformat)->Bmask = 0;
    (*pixelformat)->Amask = 0;

    //(*pixelformat)->colorkey = 0;
    //(*pixelformat)->alpha = 255;

    return (0);
}

int freepixelformat(SDL_PixelFormat **pixelformat){
    free ((*pixelformat)->palette->colors);
    free ((*pixelformat)->palette);
    free (*pixelformat);
    return (0);
}
